Add support for generic player chat commands (trigger words)

cemathey opened the pull request:
* Add generic player chat commands (trigger words)
* Add a cache (`MostRecentEvents`) that is updated every kill or team kill 
* Refactored scorebot/scoreboard stuff a little bit to use their types/lookups elsewhere
* Adds a `discord_invite_url` option to `RconServerSettingsType`
* Updates the webhook user configs to return a sorted list of admin ping words to guarantee the order when running tests
* Adds `SafeStringFormat` a subclass of `dict` that will accept any formatted string and replace missing keys with the key itself to avoid exceptions when using `"".format_map()"`

This supports two types of information, info we can get from anywhere in the RCON without any extra details (server name, etc.) and info that is only available depending on the context. This is represented by two types (`MessageVariable` and `MessageVariableContext`).

It does pass in the `steam_id_64` when used in `rcon.hooks.trigger_words` since the chatting players steam ID is always available and allows us to look up VIP status, etc.

Users can add as many commands/command formats as they want, but they can only add `{variable}` style variables if they're either in the `MessageVariable` or `MessageVariableContext` enums to avoid typos, etc.

This does require some more work to support the rest of the scoreboard types, and if this is an acceptable way to do this, we could use this across the board in other places in RCON (automods, etc.) to allow people more options when automatically messaging players.

We can also use this to add end of round stuff for #392

ElGuillermo commented:
Seems quite good for me, but I'd prefer FlorianSW to approve this one.

FlorianSW commented on rcon/hooks.py:
```diff
@@ -76,6 +89,97 @@ def initialise_vote_map(rcon: Rcon, struct_log):
+        else MostRecentEvents()
+    )
+
+    chat_words = set(re.split(CHAT_WORDS_RE, chat_message))
```
The second thing I have trouble to understand, tbh 😅 
I _think_ the only reason of this code is to separate words based to be able to look for trigger words, right? What I don't get is: Why would we separate `!who-killed-me` to `["!who", "killed", "me"]`? Shouldn't it be one entry in the array as the whole group of words? Like only splitting by spaces? It would be as simple as:
```
chat_words = chat_message.split(' ')
```

After all: why would you need to split the chat words at all? Wouldn't a simple "contains" for the trigger.words be sufficient? 🤔

cemathey commented on rconweb/api/user_settings.py:
```diff
@@ -2474,6 +2474,85 @@ def set_watchlist_discord_webhooks_config(request):
 
+@csrf_exempt
+@login_required()
+# TODO permission
```
Whoops, yeah I need to add that, good catch.

cemathey commented on tests/test_message_variables.py:
```diff
@@ -0,0 +1,282 @@
+    monkeypatch.setattr(
+        rcon.hooks,
+        "get_recent_actions",
+        lambda x: {steam_id_64: recent_event},
```
Yeah whoops, this was out of date from before I split the cache into two calls, one to set and one to get.

cemathey commented on tests/test_message_variables.py:
```diff
@@ -0,0 +1,282 @@
+            ),
+            "1234",
+            MostRecentEvents(
+                last_victim_name="some guy", last_victim_weapon="some weapon"
```
Out of date from when I had them merged, then split them for reduced cache size, but we decided to put them back for ease of use.

cemathey commented:
Okay I:

* Added the player names back into the cache so it's easier to use
* Expanded/updated the tests
* Renamed everything so that these are `command words` instead of `trigger words`, it's more descriptive and it won't conflict with names if we add a generic BM style trigger system later
* Forced commands to be limited to `!` and `@` prefixes only
* Set `?` as the prefix to trigger the description of a command
* Set the recent player actions (kills, etc.) cache to be reset on match start
* Updated the recent player actions cache TTL to 2.5 hours since that's the longest possible HLL match currently
* Forced the player cache to update the redis cache when it's set so it's available in other Python interpreter instances
* Added the permissions for the endpoints

FlorianSW commented on rcon/recent_actions.py:
```diff
@@ -79,14 +88,16 @@ def update_tks(rcon: Rcon, log: StructuredLogLineWithMetaData):
 
     set_recent_actions(cached_actions)
+    # Refresh the redis cache so it's available for interprocess communication
+    get_recent_actions()
```
Sounds like premature optimization to me :D But it's ok.

cemathey replied:
It's actually required if any other process wants to be able to pull it out of the redis cache, it clears on setting, and any other process that was to call it would just get back a totally empty one.

I don't know if we even want it available in any other process, but if another service wants those values for any reason it has to be done like this (or persisted to the database)

FlorianSW commented:
Looks good :) Feel free to merge after the conflicts are resolved 🤗
